### DISCUSSION TITLE
which is better way to automate fluxcd install bootstrap, python/go script or github actions

### DISCUSSION THREAD

**Discussion by kbaisla31** [fluxcd/flux2#2420] [General]
I am looking into ways to automate fluxcd install and bootstrap and would like to know the best approach to follow. We are planning to host Jenkins using fluxcd in different accounts and would like to use fluxcd multi tenancy. Any suggestions would be helpful

**Comment by kingdonb** [comment]
The Bootstrap process is something that is intended to do once per cluster, whether you are using Flux in single or multi-tenant mode (tell about cluster-wide resources and mention something about how a tenant is restricted). Is the goal to have ephemeral environments for pull requests, or something else? You may want to use ephemeral clusters and manage them remotely, with a Flux management cluster, instead of bootstrapping each cluster as it comes up. Or you may be running independent Jenkins services for each dev-team tenant. It isn't clear to me from your description of the problem.

We need more information to answer this question. We would like to help answer the question but can you describe your architecture in some depth first, to help us better understand where you're stuck?

**Reply by kbaisla31** [reply]
I am good on the bootstrap process but now I have question regarding multi tenancy. Don't we need to use seperate branches for each environment for example, dev branch for dev, staging for non-prod and main for prod clusters? I already looked at flux2 multitenancy document and it wasn't clear from the documentation.

**Reply by kingdonb** [reply]
No, there is no branching required in the Flux model described by our examples. The multi-tenancy example uses only one branch, and the [multi-environment example](https://github.com/fluxcd/flux2-kustomize-helm-example) also uses only one branch. Combining this with the multi-tenancy model you get the mechanism through Kustomize to utilize apps from a common base with minimal repetition between the environments (you can have patches that only apply in Development or Staging for example.)

As the GitRepository source has to refer to a branch declaratively, there is no elegant way to change the branch reference when branching (or to change it back when merging from the branch.) So these examples all consider that many clusters, with many apps, many environments, and even many tenants, can all be operated from one Git repository with one effective branch.

**Reply by kbaisla31** [reply]
Thanks for explaining that. One more thing which is not clear to me is, Can we have tenant and platform admin repository on the same repo as stated in the example, I see you have tenant repo on dev-team branch or we could have them on same branch. What is the recommended way to do it if you have multiple tenants?

**Reply by kingdonb** [reply]
This is a good question @kbaisla31 

If you have multiple tenants then you need a way to manage their access to write to the git repository. If we are not using environment branches then protected branches or permissioned branches are not viable strategies, even if that was supported by the git host (which GitHub out of those two hypothetical strategies only supports "protected branches" not "permissioned branches" or ACLs as far as I can tell.)

If your repo structure is such that multiple tenants are in the same repository, then you need another scheme to enforce separation. If your tenants have their own applications which are developed in separate repositories, and they do this development independently of a central authority, the repository is a natural edge/border on which that enforcement is absolutely possible. Team A does their dev in Repo A and Team B does their dev in Repo B, so nobody can step on anyone else's toes on purpose or accidentally.

Q: "Branches must be approved before they can be merged into the protected branch, so is this really necessary?" A: This is another reasonable question for monorepo users, but if you are using a mono-repository, why? Micro-services are not monoliths and Flux's strength is as a microservice which is designed to support many teams that are developing many microservices. Those efforts should be independent and free-standing.

You do not want a structure which allows two members of Team A to override decisions in Team B's area simply because there is an approver and two people agreed. Maybe you do have a system like CODEOWNERS in place to prevent this, and that's fine. Flux is not prescriptive about how tenant roles are enforced in the repository. Maybe you have a small team and good backups, and everyone is trusted to follow instructions. That's also fine!

The "dev-team" branch in the example is probably a different repository in reality. It can be a branch on the same repo, I think it was done in the example to keep things self-contained, but you should note the branch in the example is not intended to be merged back to the main branch. Maybe an explanation of this is warranted in the doc as well.
